cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -414,30 +522,50 @@ public boolean apply(RunRecordMeta input) {
     }
   }
 
-  private Map<ProgramRunId, RunRecordMeta> getSuspendedRuns(ProgramId programId, long startTime, long endTime,
+  private Map<ProgramRunId, RunRecordMeta> getSuspendedRuns(@Nullable ProgramId programId, long startTime, long endTime,
                                                              int limit, @Nullable Predicate<RunRecordMeta> filter) {
     return getNonCompleteRuns(programId, TYPE_RUN_RECORD_SUSPENDED, startTime, endTime, limit, filter);
   }
 
-  private Map<ProgramRunId, RunRecordMeta> getActiveRuns(ProgramId programId, final long startTime, final long endTime,
-                                                          int limit, @Nullable Predicate<RunRecordMeta> filter) {
+  private Map<ProgramRunId, RunRecordMeta> getActiveRuns(@Nullable ProgramId programId, final long startTime,
+                                                         final long endTime, int limit,
+                                                         @Nullable Predicate<RunRecordMeta> filter) {
     return getNonCompleteRuns(programId, TYPE_RUN_RECORD_STARTED, startTime, endTime, limit, filter);
   }
 
-  private Map<ProgramRunId, RunRecordMeta> getNonCompleteRuns(ProgramId programId, String recordType,
-                                                 final long startTime, final long endTime, int limit,
-                                                 Predicate<RunRecordMeta> filter) {
+  private Map<ProgramRunId, RunRecordMeta> getNonCompleteRuns(@Nullable ProgramId programId, String recordType,
+                                                              final long startTime, final long endTime, int limit,
+                                                              Predicate<RunRecordMeta> filter) {
     MDSKey activeKey = getProgramKeyBuilder(recordType, programId).build();
+    Predicate<RunRecordMeta> finalPredicate = andPredicate(new Predicate<RunRecordMeta>() {
 
-    Map<MDSKey, RunRecordMeta> runRecordMap = listKV(activeKey, null, RunRecordMeta.class, limit,
-                                                     andPredicate(new Predicate<RunRecordMeta>() {
       @Override
       public boolean apply(RunRecordMeta input) {
         return input.getStartTs() >= startTime && input.getStartTs() < endTime;
       }
-    }, filter));
-
-    return getProgramRunIdMap(runRecordMap);
+    }, filter);
+
+    Map<MDSKey, RunRecordMeta> runRecordMap = listKV(activeKey, null, RunRecordMeta.class, limit, finalPredicate);
+    Map<ProgramRunId, RunRecordMeta> programRunIdRunRecordMetaMap = getProgramRunIdMap(runRecordMap);
+
+    int remaining = limit - programRunIdRunRecordMetaMap.size();
+    if (programId != null && programId.getVersion().equals(ApplicationId.DEFAULT_VERSION) && (remaining > 0)) {
+      activeKey = getVersionLessProgramKeyBuilder(recordType, programId).build();
+      runRecordMap = listKV(activeKey, null, RunRecordMeta.class, remaining, finalPredicate);
```

Comment:
Instead of running two scans for every request, we can change the scan when programId is default version. In that case do a versionless scan and add a filter to return only records with the default version.